README
This is Perl module Catalyst::Engine::PSGI.

INSTALLATION

Catalyst::Engine::PSGI installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Catalyst::Engine::PSGI

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Catalyst::Engine::PSGI documentation is available as in POD. So you can do:

    % perldoc Catalyst::Engine::PSGI

to read the documentation online with your favorite pager.

Tatsuhiko Miyagawa


Changes
Revision history for Perl extension Catalyst::Engine::PSGI

{{$NEXT}}

0.14  2015-04-19 11:08:06 CEST
        - Deprecate this module

0.13  Thu Jun  9 23:58:10 PDT 2011
        - Fixed a bug in PSGI header generation to prevent potential HTTP header injection
        - Misc doc fixes

0.12  Thu Jan  6 14:37:53 PST 2011
        - Fix for Catalyst::Runtime >= 5.80030 (pedromelo)

0.11  Fri Jul 30 12:49:46 PDT 2010
        - Allows setting a code reference to the $c->res->body as
          a raw PSGI streaming (t0m)

0.10  Thu May 20 14:38:03 PDT 2010
        - Added docs for frontend proxy and X-Forwarded-* headers

0.09  Wed Apr  7 20:17:57 PDT 2010
        - Escape :: in the helpers for .psgi files (lestrrat)

0.08  Fri Mar 19 00:58:35 PDT 2010
        - Fixed a problem with URLMap where root access gets an empty path (omega)

0.07  Thu Dec 10 22:14:04 PST 2009
        - Clear env reference to avoid leaks in Servers like AnyEvent (chiba)
        - Added Catalyst::Controller::Metal

0.06  Sat Nov  7 20:20:45 PST 2009
        - Added Action::RenderView dependency

0.05  Mon Oct 26 00:44:11 PDT 2009
        - Removed another Catalyst dev dependency

0.04  Mon Oct 19 19:41:51 PDT 2009
        - Bump up Catalyst dep to 5.8 (thanks to mst)

0.03  Fri Oct 16 00:21:37 PDT 2009
        - Added Test::Requires

0.02  Tue Oct 13 13:28:49 PDT 2009
        - Do not use Static::Simple in the test (Thanks to beppu)

0.01  Sat Sep  5 17:29:03 2009
        - original version


t/Hello/Makefile.PL
#!/usr/bin/env perl
# IMPORTANT: if you delete this file your app will not work as
# expected.  You have been warned.
use inc::Module::Install;

name 'Hello';
all_from 'lib/Hello.pm';

requires 'Catalyst::Runtime' => '5.80007';
requires 'Catalyst::Plugin::ConfigLoader';
requires 'Catalyst::Plugin::Static::Simple';
requires 'Catalyst::Action::RenderView';
requires 'parent';
requires 'Config::General'; # This should reflect the config file format you've chosen
                 # See Catalyst::Plugin::ConfigLoader for supported formats
catalyst;

install_script glob('script/*.pl');
auto_install;
WriteAll;


t/Hello/lib/Hello.pm
package Hello;

use strict;
use warnings;

use Catalyst::Runtime 5.80;

# Set flags and add plugins for the application
#
#         -Debug: activates the debug mode for very useful log messages
#   ConfigLoader: will load the configuration from a Config::General file in the
#                 application's home directory
# Static::Simple: will serve static files from the application's root
#                 directory

use parent qw/Catalyst/;
use Catalyst qw/-Debug/;
our $VERSION = '0.01';

# Configure the application.
#
# Note that settings in hello.conf (or other external
# configuration file that you set up manually) take precedence
# over this when using ConfigLoader. Thus configuration
# details given here can function as a default configuration,
# with an external configuration file acting as an override for
# local deployment.

__PACKAGE__->config( name => 'Hello' );

# Start the application
__PACKAGE__->setup();


=head1 NAME

Hello - Catalyst based application

=head1 SYNOPSIS

    script/hello_server.pl

=head1 DESCRIPTION

[enter your description here]

=head1 SEE ALSO

L<Hello::Controller::Root>, L<Catalyst>

=head1 AUTHOR

miyagawa

=head1 LICENSE

This library is free software. You can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


t/Hello/script/hello_cgi.pl
#!/usr/bin/env perl

BEGIN { $ENV{CATALYST_ENGINE} ||= 'CGI' }

use strict;
use warnings;
use FindBin;
use lib "$FindBin::Bin/../lib";
use Hello;

Hello->run;

1;

=head1 NAME

hello_cgi.pl - Catalyst CGI

=head1 SYNOPSIS

See L<Catalyst::Manual>

=head1 DESCRIPTION

Run a Catalyst application as a cgi script.

=head1 AUTHORS

Catalyst Contributors, see Catalyst.pm

=head1 COPYRIGHT


This library is free software. You can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Catalyst/Controller/Metal.pm
package Catalyst::Controller::Metal;
use Moose;
extends 'Catalyst::Component';
with 'Catalyst::Component::ApplicationAttribute';

my %metals;

sub BUILD {
    my($self, $args) = @_;
    push @{$metals{$self->_application}}, $self;
}

sub metals_for {
    my($class, $application) = @_;
    @{$metals{$application} || []};
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

Catalyst::Controller::Metal - Raw PSGI handling in Catalyst controllers

=head1 SYNOPSIS

  package MyApp::Controller::Metalic;
  use parent 'Catalyst::Controller::Metal';

  sub call {
      my($self, $env) = @_;
      if ($env->{PATH_INFO} =~ m!^/hello!) {
          return [ 200, [ "Content-Type" => 'text/plain' ], [ "Hello World" ] ];
      } else {
          return [ 404, [], [] ];
      }
  }

Catalyst::Controller::Metal allows you to write a raw PSGI handler in
your Catalyst application, inspired by Rails Metal that allows you to
write raw Rack applications inside Ruby on Rails.

=head1 SEE ALSO

Rails Metal.

=cut


lib/Catalyst/Engine/PSGI.pm
package Catalyst::Engine::PSGI;

use strict;
use 5.008_001;
our $VERSION = '0.14';

use Moose;
extends 'Catalyst::Engine';

{
    # Temporary hack to see if there are better ways like TraitFor,
    # but without requiring downstream changes.
    sub Catalyst::Request::env {
        my $req = shift;
        $req->{_psgi_env} = shift if @_;
        $req->{_psgi_env};
    }
}

use Scalar::Util qw(blessed);
use URI;
use Catalyst::Controller::Metal;

# This is what Catalyst does to decode path. Not compatible to CGI RFC 3875
my %reserved = map { sprintf('%02x', ord($_)) => 1 } split //, $URI::reserved;
sub _uri_safe_unescape {
    my ($s) = @_;
    $s =~ s/%([a-fA-F0-9]{2})/$reserved{lc($1)} ? "%$1" : pack('C', hex($1))/ge;
    $s
}

sub prepare_connection {
    my ( $self, $c ) = @_;

    my $request = $c->request;
    my $env = $self->env;

    $request->env($env);
    $request->address( $env->{REMOTE_ADDR} );
    $request->hostname( $env->{REMOTE_HOST} ) if exists $env->{REMOTE_HOST};
    $request->protocol( $env->{SERVER_PROTOCOL} );
    $request->user( $env->{REMOTE_USER} );  # XXX: Deprecated. See Catalyst::Request for removal information
    $request->remote_user( $env->{REMOTE_USER} );
    $request->method( $env->{REQUEST_METHOD} );

    $request->secure( $env->{'psgi.url_scheme'} eq 'https' );
}

sub prepare_headers {
    my ( $self, $c ) = @_;

    my $env = $c->request->env;
    my $headers = $c->request->headers;
    foreach my $header ( keys %$env ) {
        next unless $header =~ /^(HTTP|CONTENT|COOKIE)/i;
        ( my $field = $header ) =~ s/^HTTPS?_//;
        $field =~ tr/_/-/;
        $headers->header( $field => $env->{$header} );
    }
}

sub prepare_path {
    my ( $self, $c ) = @_;

    my $env = $c->request->env;

    my $scheme = $c->request->secure ? 'https' : 'http';
    my $host      = $env->{HTTP_HOST} || $env->{SERVER_NAME};
    my $port      = $env->{SERVER_PORT} || 80;
    my $base_path = $env->{SCRIPT_NAME} || "/";

    # set the request URI
    my $req_uri = $env->{REQUEST_URI};
       $req_uri =~ s/\?.*$//;
    my $path = _uri_safe_unescape($req_uri);
    if ($path eq $base_path) {
        $path .= "/"; # To fool catalyst a bit
    }
    $path =~ s{^/+}{};

    # Using URI directly is way too slow, so we construct the URLs manually
    my $uri_class = "URI::$scheme";

    # HTTP_HOST will include the port even if it's 80/443
    $host =~ s/:(?:80|443)$//;

    if ( $port !~ /^(?:80|443)$/ && $host !~ /:/ ) {
        $host .= ":$port";
    }

    # Escape the path
    $path =~ s/([^$URI::uric])/$URI::Escape::escapes{$1}/go;
    $path =~ s/\?/%3F/g; # STUPID STUPID SPECIAL CASE

    my $query = $env->{QUERY_STRING} ? '?' . $env->{QUERY_STRING} : '';
    my $uri   = $scheme . '://' . $host . '/' . $path . $query;

    $c->request->uri( bless \$uri, $uri_class );

    # set the base URI
    # base must end in a slash
    $base_path .= '/' unless $base_path =~ m{/$};

    my $base_uri = $scheme . '://' . $host . $base_path;

    $c->request->base( bless \$base_uri, $uri_class );
}

around prepare_query_parameters => sub {
    my $orig = shift;
    my ( $self, $c ) = @_;

    if ( my $qs = $c->request->env->{QUERY_STRING} ) {
        $self->$orig( $c, $qs );
    }
};

sub prepare_request {
    my ( $self, $c, %args ) = @_;

    if ( $args{env} ) {
        $self->env( $args{env} );
    }

    $self->{buffer} = '';
}

sub write {
    my($self, $c, $buffer) = @_;
    $self->{buffer} .= $buffer if defined $buffer;
}

sub finalize_body {
    # do nothing since we serve content
}

sub read_chunk {
    my($self, $c) = (shift, shift);
    $self->env->{'psgi.input'}->read(@_);
}

sub run {
    my($self, $class, $env) = @_;

    # short circuit with Metal
    for my $metal (Catalyst::Controller::Metal->metals_for($class)) {
        my $res = $metal->call($env);
        if (defined $res && !(ref $res eq 'ARRAY' && $res->[0] == 404)) {
            return $res;
        }
    }

    # what Catalyst->handle_request does
    my $status = -1;
    my $c;
    eval {
        $c = $class->prepare(env => $env);
        $c->dispatch;
        $status = $c->finalize;
    };

    # clear the $env ref to avoid leaks
    $self->env(undef);

    if (my $error = $@) {
        chomp $error;
        $class->log->error(qq/Caught exception in engine "$error"/);
    }

    if (my $coderef = $class->log->can('_flush')){
        $class->log->$coderef();
    }

    return [ 500, [ 'Content-Type' => 'text/plain', 'Content-Length' => 11 ], [ 'Bad request' ] ]
        unless $c;

    my $body = $c->res->body;
    $body = '' unless defined $body;
    if (!ref $body && $body eq '' && $self->{buffer}) {
        $body = [ $self->{buffer} ];
    } elsif (ref($body) eq 'GLOB' || (Scalar::Util::blessed($body) && $body->can('getline'))) {
        # $body is FH
    } elsif (ref($body) eq 'CODE') {
        return $body;
    } else {
        $body = [ $body ];
    }

    my $headers = [];
    $c->res->headers->scan(sub { my($k, $v) = @_; push @$headers, $k, _escape($v) });
    return [ $c->res->status, $headers, $body ];
}

sub _escape {
    local $_ = shift;
    s/(\r?\n)+/ /g;
    return $_;
}

no Moose;

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Catalyst::Engine::PSGI - PSGI engine for Catalyst

=head1 WARNINGS

B<Catalyst 5.9000 or later has a built-in PSGI support and this module is deprecated.>

=head1 SYNOPSIS

  # app.psgi
  use strict;
  use MyApp;

  MyApp->setup_engine('PSGI');
  my $app = sub { MyApp->run(@_) };

=head1 DESCRIPTION

Catalyst::Engine::PSGI is a Catalyst Engine that adapts Catalyst into the PSGI gateway protocol.

=head1 COMPATIBILITY

=over 4

=item *

Currently this engine works with Catalyst 5.8 (Catamoose) or newer.

=item *

Your application is supposed to work with any PSGI servers without any
code modifications, but if your application uses C<< $c->res->write >>
to do streaming write, this engine will buffer the ouput until your
app finishes.

To do real streaming with this engine, you should implement an
IO::Handle-like object that responds to C<getline> method that returns
chunk or undef when done, and set that object to C<< $c->res->body >>.

Alternatively, it is possible to set the body to a code reference,
which will be used to stream content as documented in the
L<PSGI spec|PSGI/Delayed_Reponse_and_Streaming_Body>.

=item *

When your application runs behind the frontend proxy like nginx or
lighttpd, this Catalyst engine doesn't automatically recognize the
incoming headers like C<X-Forwarded-For>, because respecting these
headers by default causes a potential security issue.

You have to enable L<Plack::Middleware::ReverseProxy> or
L<Plack::Middleware::ForwardedHeaders> to automatically promote those
forwarded headers into C<REMOTE_ADDR> hence IP address of the request.

ReverseProxy middleware is pretty simple and has no configuration
while ForwardedHeaders allows you to configure which upstream host to
trust, etc.

=back

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

Most of the code is taken and modified from Catalyst::Engine::CGI.

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

I<Catalyst::Engine> L<PSGI> I<Plack>

=cut
